Update remix v2 doc (#9317)
marmelab/react-admin: docs/Remix.md

## Remix v2

### Setting Up Remix

Let's start by creating a new Remix project. Run the following command:

```sh
npx create-remix@latest
```

This script will ask you for more details about your project. You can use the following options:

- The name you want to give to your project, e.g. `remix-admin`
- Initialize a new git repository? Choose Yes
- Install dependencies with npm? Choose Yes

> The project structure should look something like this:  

Reviewer: ```suggestion
The project structure should look like this:  
```

![Remix v2 project structure](./img/remix-v2-structure.png)

### Setting Up React-Admin In Remix

Add the `react-admin` npm package, as well as a data provider package. In this example, we'll use `ra-data-json-server` to connect to a test API provided by [JSONPlaceholder](https://jsonplaceholder.typicode.com).

```sh
cd remix-admin
npm add react-admin ra-data-json-server
```

> **Tip**: If you're using yarn, Remix and react-admin both install `react-router`, and due to the way each library handles its dependency, this results in duplicate packages. To avoid this, use [yarn resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) to force Remix to use the same version of `react-router` as react-admin. So add the following to the `package.json` file:

Reviewer: ```suggestion
**Tip**: If you're using yarn, Remix and react-admin both install `react-router`, and due to the way each library handles its dependencies, this results in duplicate packages. To avoid this, use [yarn resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) to force Remix to use the same version of `react-router` as react-admin. So add the following to the `package.json` file:
```

```js
{
  // ...
  "resolutions": {
    "react-router": "6.8.1",
    "react-router-dom": "6.8.1"
  }
}
```

And call `yarn` again to install the dependencies:

```sh
yarn
```

### Using React-Admin As The Root Application

> Remix and react-admin both use [react-router](https://reactrouter.com/) for routing. React-admin detects when it is included inside an existing React Router context and reuses it. This is problematic because Remix uses file-based routing. So when react-admin changes the route to `/posts` for instance, Remix will look for a corresponding `app/routes/posts.tsx` file. As it doesn't exist, Remix will render a 404.

Reviewer: ```suggestion
Remix and react-admin both use [react-router](https://reactrouter.com/) for routing. React-admin detects when it is included itself inside an existing React Router context. If so, react-admin will reuses it. This is problematic because Remix uses file-based routing. So when react-admin changes the route to `/posts` for instance, Remix will look for a corresponding `app/routes/posts.tsx` file. As it doesn't exist, Remix will render a 404.
```

Author: ```suggestion
Remix and react-admin both use [react-router](https://reactrouter.com/) for routing. React-admin detects when it is included inside an existing React Router context. If so, react-admin will reuses it. This is problematic because Remix uses file-based routing. So when react-admin changes the route to `/posts` for instance, Remix will look for a corresponding `app/routes/posts.tsx` file. As it doesn't exist, Remix will render a 404.
```

> Remix and react-admin both use [react-router](https://reactrouter.com/) for routing. React-admin detects when it is included inside an existing React Router context. If so, react-admin will reuses it. This is problematic because Remix uses file-based routing. So when react-admin changes the route to `/posts` for instance, Remix will look for a corresponding `app/routes/posts.tsx` file. As it doesn't exist, Remix will render a 404.

Reviewer 2: ```suggestion
Remix and react-admin both use [react-router](https://reactrouter.com/) for routing. React-admin detects when it is included inside an existing React Router context. If so, react-admin will reuse it. This is problematic because Remix uses file-based routing. So when react-admin changes the route to `/posts` for instance, Remix will look for a corresponding `app/routes/posts.tsx` file. As it doesn't exist, Remix will render a 404.
```

body { margin: 0; }
```

> **Tip**: Remix doesn't let splat routes catch requests to the index page ('/'), so you must have both the `app/routes/index.tsx` and `app/routes/$.tsx` routes to correctly render the admin app.

Reviewer: ```suggestion
**Tip**: Remix doesn't let splat routes catch requests to the index page ('/'), so you need to have both the `app/routes/index.tsx` and `app/routes/$.tsx` files to correctly render the admin app.
```

> **Tip**: Remix doesn't let splat routes catch requests to the index page ('/'), so you need to have both the `app/routes/index.tsx` and `app/routes/$.tsx` files to correctly render the admin app.

Reviewer 2: This is duplicate to what's said in the beginning of the paragraph. Besides, it's `app/routes/_index.tsx` and not `app/routes/index.tsx`.

Next, create the admin app component in `app/components/App.tsx`:

```jsx
// in app/components/App.tsx
import { Admin, Resource, ListGuesser } from "react-admin";
import jsonServerProvider from "ra-data-json-server";

const dataProvider = jsonServerProvider("https://jsonplaceholder.typicode.com");

const App = () => (
  <Admin dataProvider={dataProvider}>
    <Resource name="posts" list={ListGuesser} />
    <Resource name="comments" list={ListGuesser} />
  </Admin>
);

export default App;
```

> This is a minimal admin for 2 resources. React-admin should be able to render a list of posts and a list of comments, guessing the data structure from the API response. 

Reviewer: ```suggestion
This is a minimal admin for two resources. React-admin should be able to render a list of posts and a list of comments, guessing the data structure from the API response. 
```

> Now, start the server with `npm run dev`, browse to `http://localhost:3000/`, and you should see the working admin:

Reviewer 2: I have the following error at this stage:

```
 💿  remix dev

 info  building...
 info  built (6.7s)
TypeError: jsonServerProvider is not a function
    at file:///home/slax57/workspaces/ra-test-remix-v2/build/index.js?t=1696254116269.5742:205:20
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)
```

In many cases, the admin is only a part of the application. For instance, you may want to render the admin in a subpath like `/admin`.

To do so, add a [splat route](https://remix.run/docs/en/main/file-conventions/routes#splat-routes), i.e. a route that matches all URLs inside a sub path. A splat route is named `$.tsx`. Create a file called `app/routes/admin.$.tsx` file with the following content:

```jsx
// in app/routes/admin.$.tsx
import { Admin, Resource, ListGuesser } from "react-admin";
import jsonServerProvider from "ra-data-json-server";
import styles from "~/styles/app.css";

export function links() {
  return [{ rel: "stylesheet", href: styles }];
}

> const dataProvider = jsonServerProvider.default

Reviewer 2: I have the following TS warning at this line:

```
Property 'default' does not exist on type '(apiUrl: any, httpClient?: ((url: any, options?: Options | undefined) => Promise<{ status: number; headers: Headers; body: string; json: any; }>) | undefined) => DataProvider'.ts(2339)
```

  ? jsonServerProvider.default("https://jsonplaceholder.typicode.com")
  : jsonServerProvider("https://jsonplaceholder.typicode.com");

export default function App() {
  return (
    <Admin basename="/admin" dataProvider={dataProvider}>
      <Resource name="posts" list={ListGuesser} />
      <Resource name="comments" list={ListGuesser} />
    </Admin>
  );

> }

Reviewer 2: Why not reuse the `App` component here? Following the tutorial, we now have a useless `App.tsx` file...

Besides, this file is still referenced in the following paragraph, so you should definitely keep it!

Author: In fact, I think we should remove the "Using React-Admin As The Root Application" altogether

body { margin: 0; }
```

**Tip** Don't forget to set the `<Admin basename>` prop, so that react-admin generates links relative to the "/admin" subpath:

> Now the admin renders at `http://localhost:3000/admin`, and you can use the Remix routing system to add more pages.

Reviewer 2: I have the following error at this stage

```
 💿  remix dev

 info  building...
 info  built (6.2s)
TypeError: jsonServerProvider2 is not a function
    at file:///home/slax57/workspaces/ra-test-remix-v2/build/index.js?t=1696254416879.3103:242:21
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)
```

};
```

> **Tip**: Some of this code is really PostgREST-specific. The `prefer` header is required to let PostgREST returns one record instead of an array containing one record in response to `getOne` requests. A proxy for another CRUD API will require different parameters.

Reviewer 2: Nope, grammar was good before.

```suggestion
**Tip**: Some of this code is really PostgREST-specific. The `prefer` header is required to let PostgREST return one record instead of an array containing one record in response to `getOne` requests. A proxy for another CRUD API will require different parameters.
```
